jolt-core/src/r1cs/builder.rs

```diff
@@ -0,0 +1,1443 @@
+    input_to_flat: Vec<Option<Range<usize>>>,
+    compute: Box<AuxComputationFunction<F>>,
+}
+
+impl<F: JoltField, I: ConstraintInput> AuxComputation<F, I> {
+    fn new(
+        output: Variable<I>,
+        symbolic_inputs: Vec<LC<I>>,
+        compute: Box<AuxComputationFunction<F>>,
+    ) -> Self {
+        let flat_var_count: usize = symbolic_inputs.iter().map(|input| input.num_vars()).sum();
+        let mut flat_vars = Vec::with_capacity(flat_var_count);
+        let mut input_to_flat = Vec::with_capacity(symbolic_inputs.len());
+
+        let mut range_start_index = 0;
+        for input in &symbolic_inputs {
+            let terms = input.terms();
+            let num_vars = input.num_vars();
+            for term in terms {
+                if let Variable::Constant = term.0 {
+                    continue;
+                }
+                flat_vars.push(term.0);
+            }
+            if num_vars > 0 {
+                input_to_flat.push(Some(range_start_index..(range_start_index + num_vars)));
+                range_start_index += num_vars;
+            } else {
+                input_to_flat.push(None);
+            }
+        }
+        assert_eq!(flat_vars.len(), flat_var_count);
+
+        #[cfg(test)]
+        {
+            let output_index = match output {
+                Variable::Auxiliary(output_index) => output_index,
+                _ => panic!("Output must be of the Variable::Aux variant"),
+            };
+            for aux_var in &flat_vars {
+                if let Variable::Auxiliary(aux_calc_index) = aux_var {
+                    // Currently do not support aux computations dependent on those allocated after. Could support with dependency graph, instead
+                    // dev should write their constraints sequentially. Simplifies aux computation parallelism.
+                    if output_index <= *aux_calc_index {
+                        panic!("Aux computation depends on future aux computation: {output:?} = f({aux_var:?})");
+                    }
+                }
+            }
+        }
+
+        Self {
+            output,
+            symbolic_inputs,
+            flat_vars,
+            input_to_flat,
+            compute,
+        }
+    }
+
+    /// Takes one value per value in flat_vars.
+    fn compute(&self, values: &[F]) -> F {
+        assert_eq!(values.len(), self.flat_vars.len());
+        assert_eq!(self.input_to_flat.len(), self.symbolic_inputs.len());
+        let computed_inputs: Vec<_> = self
+            .symbolic_inputs
+            .iter()
+            .enumerate()
+            .map(|(input_index, input_lc)| {
+                let values = if let Some(range) = self.input_to_flat[input_index].clone() {
+                    &values[range]
+                } else {
+                    &[]
+                };
+                input_lc.evaluate(values)
+            })
+            .collect();
+        (self.compute)(&computed_inputs)
+    }
```

Comment:
it might be annoying to get this to work with the Rust compiler, but it would be cool if, instead of having this as a method, we generate this as a closure in `new` and store it as the `compute` field in `AuxComputation`. Then we could get rid of the `input_to_flat` field I think

Reply:
Alternatively, we could pass in all the inputs/aux values for a given step into this function. I'm not a huge fan of having `flat_vars` and `input_to_flat` as fields in the `AuxComputation` struct.

Reply:
These have been removed as of #384